### ISSUE TITLE
Improve help() by making typing.overload() information accessible at runtime

### ISSUE THREAD

**Issue by rhettinger** [python/cpython#89263]
BPO | [45100](https://bugs.python.org/issue45100)
--- | :---
Nosy | @gvanrossum, @rhettinger, @ronaldoussoren, @JelleZijlstra, @TeamSpen210, @sobolevn, @Fidget-Spinner, @AlexWaygood, @DiddiLeija
PRs | <li>python/cpython#31716</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2021-09-04.18:03:28.637>
labels = ['3.11']
title = 'Improve help() by making typing.overload() information accessible at runtime'
updated_at = <Date 2022-03-28.22:36:21.459>
user = 'https://github.com/rhettinger'
```

bugs.python.org fields:
```python
activity = <Date 2022-03-28.22:36:21.459>
actor = 'gvanrossum'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = []
creation = <Date 2021-09-04.18:03:28.637>
creator = 'rhettinger'
dependencies = []
files = []
hgrepos = []
issue_num = 45100
keywords = ['patch']
message_count = 14.0
messages = ['401052', '401074', '401080', '401094', '401097', '413677', '413678', '413684', '413685', '413695', '413751', '414561', '414566', '416226']
nosy_count = 9.0
nosy_names = ['gvanrossum', 'rhettinger', 'ronaldoussoren', 'JelleZijlstra', 'Spencer Brown', 'sobolevn', 'kj', 'AlexWaygood', 'DiddiLeija']
pr_nums = ['31716']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = None
url = 'https://bugs.python.org/issue45100'
versions = ['Python 3.11']
```

</p></details>

**Comment by AlexWaygood** [comment]
After #31716 is merged, I'd like to have a stab at writing a PR to have overloads shown in the output of `help()`. For now, it will probably be easiest to do that without tinkering with `inspect.signature()`, and instead only making the changes in `pydoc` -- I have a few ideas of how to do that.

**Comment by AlexWaygood** [comment]
I was just playing around with the new `get_overloads` function, and it appears that it doesn't work when using the interactive shell. This seems unfortunate -- it certainly took me by surprise. I think it also means that it's hard to use the new function to improve the output of `help()`, since `help()` has to work effectively in the interactive shell (for obvious reasons).

```python
C:\Users\alexw\coding\cpython>python
Running Debug|x64 interpreter...
Python 3.11.0a7+ (main, Apr 14 2022, 10:41:31) [MSC v.1931 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> from typing import *
>>> @overload
... def foo(arg: int) -> str: ...
...
>>> @overload
... def foo(arg: str) -> int: ...
...
>>> def foo(arg: int | str) -> str | int: ...
...
>>> get_overloads(foo)
[<function foo at 0x000001D030B0C9F0>]
>>> import inspect
>>> inspect.signature(_[0])
<Signature (arg: str) -> int>
```

**Comment by AlexWaygood** [comment]
This is because all functions defined in the interactive shell have the same line-number set on their code object:

```python
>>> def foo(): ...
...
>>> def bar(): ...
...
>>> foo.__code__.co_firstlineno == bar.__code__.co_firstlineno == 1
True
```

Maybe it's not so much of a problem, though, since most people won't be calling `help()` on functions defined _in_ the interactive shell -- they'll be calling it on functions that they _import from within_ the interactive shell.

**Comment by JelleZijlstra** [comment]
It's a little unfortunate because it makes it harder for people to experiment with `get_overloads()` in the REPL. But I agree that it's not a big problem for people using `help()` to introspect library functions.

**Comment by gvanrossum** [comment]
Since `@overload` is a static typing feature, why would you be typing it in the REPL in the first place?

**Comment by JelleZijlstra** [comment]
> Since `@overload` is a static typing feature, why would you be typing it in the REPL in the first place?

If you read about `get_overloads()` in the What's New and want to try it out for yourself, you might write some overloads in the REPL to see it in action. Looks like that's essentially what happened to Alex.

**Comment by AlexWaygood** [comment]
> Since `@overload` is a static typing feature, why would you be typing it in the REPL in the first place?

Yeah, this is why I don't think it's actually that much of a big deal, as I said above.

> If you read about `get_overloads()` in the What's New and want to try it out for yourself, you might write some overloads in the REPL to see it in action. Looks like that's essentially what happened to Alex.

Yup. Experimenting with stuff in the REPL is generally a part of my process when I write patches for CPython. I was tinkering with modifying `pydoc` to use `get_overloads`, and was trying things out in the REPL. I was surprised when things didn't work as expected.

**Comment by furkanonder** [comment]
@JelleZijlstra The issue seems to be solved. We can close the issue.

**Comment by JelleZijlstra** [comment]
We added `typing.get_overloads`, but I don't think we ended up making `help()` look at this information yet.
